src/fsharp/FSharp.Core/seq.fs

```diff
@@ -1577,57 +1577,56 @@ namespace Microsoft.FSharp.Collections
 
         [<CompiledName("Sum")>]
         let inline sum (source: seq< (^a) >) : ^a = 
-            use e = source.GetEnumerator() 
             let mutable acc = LanguagePrimitives.GenericZero< (^a) >
-            while e.MoveNext() do
-                acc <- Checked.(+) acc e.Current
+            for item in source do
+                acc <- Checked.(+) acc item
```

Comment:
if we inline *enough*, maybe the compiler can use the array for loop directly. If we ask for `.GetEnumerator` for sure not

```fsharp
let inline sum (source: seq< (^a) >) : ^a = 
```

it's not possibile to use something like

```
let inline sum (source: ^s) : ^a when ^s: seq< (^a) > 
```

so source is a `seq<(^a)>`, but can be specialized?